File: Binary Tree Zigzag Level Order Traversal/Binary_Tree_Zigzag_Level_Order_Traversal.cpp
// 4ms 63.18%

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include<vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> values;
        if (!root)return values;

        queue<TreeNode*> treeQueue;
        treeQueue.push(root);

        bool leftToRight = true;
        
        while (!treeQueue.empty())
        {
            int size = treeQueue.size();
            vector<int> tempVec;
            if (!leftToRight)
                tempVec.resize(size);

            while (size--)
            {
                TreeNode *tempNode = treeQueue.front();
                treeQueue.pop();

                if (leftToRight)
                    tempVec.push_back(tempNode->val);
                else
                    tempVec[size] = tempNode->val;

                if (tempNode->left)
                    treeQueue.push(tempNode->left);
                if (tempNode->right)
                    treeQueue.push(tempNode->right);
            }
            values.push_back(tempVec);
            leftToRight = !leftToRight;
        }
        return values;
    }
};

File: Binary Tree Zigzag Level Order Traversal/Binary_Tree_Zigzag_Level_Order_Traversal.py
# 56ms 41.18%

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        values = []
        if root is None:
            return values
        
        queue = []
        queue.append(root)

        leftToRight = True

        while len(queue) > 0:
            value = []
            size = len(queue)
            
            if not leftToRight:
                value = [0 for _ in range(size)]

            while size > 0:
                tempNode = queue.pop(0)

                if leftToRight:
                    value.append(tempNode.val)
                else:
                    value[size - 1] = tempNode.val

                if tempNode.left:
                    queue.append(tempNode.left)
                if tempNode.right:
                    queue.append(tempNode.right)

                size -= 1
            values.append(value)
            leftToRight = not leftToRight
        return values